flang/lib/Lower/IO.cpp

```diff
@@ -964,15 +964,43 @@ getFormat<Fortran::parser::PrintStmt>(
                    strTy, lenTy, labelMap, assignMap);
 }
 
-static std::tuple<mlir::Value, mlir::Value, mlir::Value>
+static std::tuple<mlir::Value, mlir::Value>
+genBufferFromExpr(Fortran::lower::AbstractConverter &converter,
+                  mlir::Location loc, const SomeExpr *e) {
+  auto &builder = converter.getFirOpBuilder();
+  if (Fortran::semantics::ExprHasTypeCategory(
+          *e, Fortran::common::TypeCategory::Character)) {
+    // Helper to query [BUFFER, LEN].
+    Fortran::lower::CharacterExprHelper helper(builder, loc);
+    auto dataLen = helper.materializeCharacter(converter.genExprValue(*e));
+    return {dataLen.first, dataLen.second};
+  }
+}
+
+/// Generate a reference to a buffer and the length of buffer.
+/// There are 3 cases - An IoUnit can be variable or
+/// a ScalarIntExpr (i.e FileUnitNumber) or a *.
```

Comment:
2 of the 3 cases are handled elsewhere. This code should only be called for internal I/O to a CHARACTER buffer, hence `genBuffer`.